### ISSUE TITLE
One small question on calculating the plane coefficients to floor reference

### ISSUE THREAD

**Issue by LT-ZHONG** [aperezyus/stairs_detection#2]
### Hi authors,
It's me again :)
I found another small question in the source code, which is the calculation of the plane coefficients to floor reference. In your code
`void Plane::getCoefficients2Floor(const Eigen::Affine3d& c2f)`
`{`
`    coefficients2f = c2f.inverse().matrix().transpose().cast<float>() * coefficients;    `
`}`
this function is used to transform the plane coefficients to floor reference.
But I think it's wired why you use `c2f.inverse()` instead of `c2f`.
`c2f.inverse()` is equal to `f2c`, isn't it? If so, it means the transformation of the plane coefficients to camera reference.
Then I modified the code, it works too. That's really strange.

I noticed the following function, which is used to transform the plane centroid to floor reference
`void Plane::getCentroid2Floor(const Eigen::Affine3d& c2f)`
In this function, you use `c2f`, not `c2f.inverse()`

**Comment by aperezyus** [comment]
Hi again!

Sorry for the delay!

I think it is in fact well done in the code. In _getCentroid2Floor_ we are transforming a point from _c_ reference frame to _f_ (hence _c2f_). But in _getCoefficients2Floor_ we are transforming **plane coefficients** instead. The way of performing this transformation is via the **inverse transpose** of the 4x4 transformation matrix (in this case, _c2f_). Here there are some discussions about this issue I've quickly found [LINK1](https://stackoverflow.com/questions/2306665/how-to-efficiently-rotate-and-translate-a-plane-in-3d), [LINK2](https://stackoverflow.com/questions/7685495/transforming-a-3d-plane-using-a-4x4-matrix), although I'm sure there are better sources out there.

In conclusion, I believe our coding was right, since I was computing the _.inverse()_ and _.transpose()_ operation to the initial transformation matrix (_c2f_) to transform the plane coefficients to the new reference. Since we had already computed _f2c_ we could've substituted _c2f.inverse()_ directly for _f2c_ for efficiency, but other than that I think the method is right.

Hope that explains the issue!

Alejandro

**Comment by LT-ZHONG** [comment]
**Hi authors,**
Thank you so much for the answer. I have fully understood why your coding is right.  The two links you provided really extend my poor math knowledge. Thank you again.

I am still researching the code and the two papers you mentioned in the github page. I noticed that 
_odometry-aided stair traversal_ was discussed in the paper [2] but it was not implemented. There are some member functions and variables appearing in the class `GlobalSceneStair`, which indicating that stair traversal module should be implemented there. So I want to finish the last part of the entire code according to your paper[2]. 
I think the procedure should be as follow.
1. Estimating the transformation matrix `Eigen::Affine3d i2s` using visual odometry
2. Using the vertical normal to correct the orientation
3. Using the edges of the steps to correct the orientation
4. Using the centroids and edges of the steps to correct the position

I have no idea if I am right or not. 
Hope that you could provide some pseudo code or source code. Any help would be greatly appreciated.
Thanks..
